symengine/refine.cpp

```diff
@@ -0,0 +1,86 @@
+#include <symengine/refine.h>
+
+namespace SymEngine
+{
+
+void RefineVisitor::bvisit(const Abs &x)
+{
+    auto farg = x.get_arg();
+    auto newarg = apply(farg);
+    if (is_true(is_nonnegative(*newarg, assumptions_))) {
+        result_ = newarg;
+    } else if (is_true(is_nonpositive(*newarg, assumptions_))) {
+        result_ = neg(newarg);
+    } else if (is_a<Conjugate>(*newarg)) {
+        result_ = abs(down_cast<const Conjugate &>(*newarg).get_arg());
+    } else {
+        result_ = abs(newarg);
+    }
+}
+
+void RefineVisitor::bvisit(const Sign &x)
+{
+    auto farg = x.get_arg();
+    auto newarg = apply(farg);
+    if (is_true(is_positive(*newarg, assumptions_))) {
+        result_ = integer(1);
+    } else if (is_true(is_negative(*newarg, assumptions_))) {
+        result_ = integer(-1);
+        //    } else if (is_true(is_zero(*newarg, assumptions_))) {
+        //        result_ = integer(0);
+    } else {
+        result_ = sign(newarg);
+    }
+}
+
+void RefineVisitor::bvisit(const Floor &x)
+{
+    auto farg = x.get_arg();
+    auto newarg = apply(farg);
+    if (is_true(is_integer(*newarg, assumptions_))) {
+        result_ = newarg;
+    } else if (could_extract_minus(*newarg)) {
+        result_ = neg(ceiling(neg(newarg)));
+    } else {
+        result_ = floor(newarg);
+    }
+}
+
+void RefineVisitor::bvisit(const Ceiling &x)
+{
+    auto farg = x.get_arg();
+    auto newarg = apply(farg);
+    if (is_true(is_integer(*newarg, assumptions_))) {
+        result_ = newarg;
+    } else if (could_extract_minus(*newarg)) {
+        result_ = neg(floor(neg(newarg)));
+    } else {
+        result_ = ceiling(newarg);
+    }
+}
+
+void RefineVisitor::bvisit(const Conjugate &x)
+{
+    auto farg = x.get_arg();
+    auto newarg = apply(farg);
+    if (is_true(is_real(*newarg, assumptions_))) {
+        result_ = newarg;
+    } else {
+        result_ = conjugate(newarg);
+    }
+}
+
+void RefineVisitor::bvisit(const UnevaluatedExpr &x)
```

Comment:
why does refine evaluate an `UnevaluatedExpr`, is that something that sympy does?

Reply:
Sympy does it in `simplify` but not in `refine`. I didn't think further than adding it because it was trivial to add. Do you think I should remove it?

Reply:
I don't know - but I would assume that having the same semantics as sympy where possible would make sense? Maybe this could be factored out into a `doit` visitor?

Reply:
Yes, that makes sense.